packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dr-response.ts

```diff
@@ -0,0 +1,202 @@
+import { XMLParser } from "fast-xml-parser";
+import dayjs from "dayjs";
+import {
+  OutboundDocumentRetrievalReq,
+  OutboundDocumentRetrievalResp,
+  XCAGateway,
+  DocumentReference,
+} from "@metriport/ihe-gateway-sdk";
+import {
+  handleRegistryErrorResponse,
+  handleHTTPErrorResponse,
+  handleEmptyResponse,
+  handleSOAPFaultResponse,
+} from "./error";
+import { parseFileFromString } from "./parse-file-from-string";
+import { S3Utils } from "../../../aws/s3";
+import { Config } from "../../../../util/config";
+import { stripUrnPrefix, constructFileName, constructFilePath } from "../utils";
+import { DRSamlClientResponse } from "../saml-client";
+
+const bucket = Config.getMedicalDocumentsBucketName();
+const region = Config.getAWSRegion();
+
+type DocumentResponse = {
+  size: string;
+  title: string;
+  creation: string;
+  language: string;
+  mimeType: string;
+  DocumentUniqueId: string;
+  HomeCommunityId: string;
+  RepositoryUniqueId: string;
+  NewDocumentUniqueId: string;
+  NewRepositoryUniqueId: string;
+  Document: string;
+};
+
+async function parseDocumentReference({
+  documentResponse,
+  outboundRequest,
+  s3Utils,
+  idMapping,
+}: {
+  documentResponse: DocumentResponse;
+  outboundRequest: OutboundDocumentRetrievalReq;
+  s3Utils: S3Utils;
+  idMapping: Record<string, string>;
+}): Promise<DocumentReference> {
+  // lets extract the document here.
+  const { mimeType, extension, decodedBytes } = parseFileFromString(documentResponse?.Document);
+  const strippedDocUniqueId = stripUrnPrefix(documentResponse?.DocumentUniqueId);
+  const metriportId = idMapping[strippedDocUniqueId];
+  if (!metriportId) {
+    throw new Error("MetriportId not found for document");
+  }
+  const fileName = constructFileName({
+    cxId: outboundRequest.cxId,
+    patientId: outboundRequest.patientId,
+    metriportId,
+    extension,
+  });
+  const filePath = constructFilePath({
+    cxId: outboundRequest.cxId,
+    patientId: outboundRequest.patientId,
+    fileName,
+  });
+  const fileInfo = await s3Utils.getFileInfoFromS3(filePath, bucket);
+
+  if (!fileInfo.exists) {
+    await s3Utils.uploadFile({
+      bucket,
+      key: filePath,
+      file: decodedBytes,
+      contentType: mimeType,
+    });
+  }
+
+  return {
+    url: "https://" + bucket + ".s3." + region + ".amazonaws.com/" + filePath,
```

Comment:
we have an existing helper for this https://github.com/metriport/metriport/blob/09611d558fb6272338d4f9e32d2c1f11823be5ab/packages/core/src/external/aws/s3.ts#L159-L161